### PR TITLE
Campaign Feature

### PR DESCRIPTION
**Pull request by jkobjacob**

### CODE REVIEW

File: contracts/campaign/Campaign.sol [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+
+    }
+
+    function createCampaign(
+        uint256 _startShuttleNum,
+        uint256 _endShuttleNum,
+        uint256 _totalRewardAmount,
+        IERC20 _rewardToken
+    ) external onlyRole(GOVERNANCE_ROLE) {
+        uint256 totalShuttles_ = _endShuttle.sub(_startShuttle);
```
Comment (cryptostaker2021) [inline contracts/campaign/Campaign.sol]: are these shuttle number inclusive? 
If I want to run for shuttle 20-40, what arguments I will pass? 
19 -40 or 20-40 ?

File: contracts/campaign/Campaign.sol [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+        emit CampaignCreated(currentCampaign, _totalRewardAmount, _startShuttleNum, _endShuttleNum, address(_rewardToken));
+    }
+
+    function claim(
+        uint256 _shuttleNumber,
+        uint256 _campaignNumber
+        uint256 _userShare,
+        address payable _sender
+    ) external {
+        uint256 rewardAmount_ = campaigns[_campaignNumber].rewardAmountPerShuttle.mul(_userShare);
```
Comment (cryptostaker2021) [inline contracts/campaign/Campaign.sol]: Add a modifier that only child pool can call this

File: contracts/campaign/Campaign.sol [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+
+contract Campaign is ICampaign, CampaignSecurityModule {
+    using SafeMath for uint256;
+
+    uint256 public currentCampaign;
+    address public owner;
+
+    mapping(uint256 => ACampaign) public campaigns;
+
+    function initialize(
```
Comment (cryptostaker2021) [inline contracts/campaign/Campaign.sol:36]: Accept child pool address

File: contracts/campaign/Campaign.sol (hunk 2 above)
Comment (cryptostaker2021) [inline contracts/campaign/Campaign.sol]: This seems wrong. 
It should be rewardAmountPerShuttle * userShare / totalShare

File: contracts/campaign/Campaign.sol [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+        owner = _owner;
+
+        _setupRole(DEFAULT_ADMIN_ROLE, _owner);
+        _setupRole(GOVERNANCE_ROLE, _owner);
+        _setupRole(CANCEL_ROLE, _owner);
+        _setupRole(PAUSE_ROLE, _owner);
+
+    }
+
+    function createCampaign(
```
Comment (cryptostaker2021) [inline contracts/campaign/Campaign.sol:63]: We need a method to delete a campaign?

**Review by cryptostaker2021 — CHANGES_REQUESTED** [review]

File: contracts/campaign/Campaign.sol [hunk 5]
```diff
@@ -67,6 +67,12 @@ contract Campaign is ICampaign, CampaignSecurityModule {
         uint256 _userShare,
         address payable _sender
     ) external onlyChildPool {
+        require(
```
Comment (jkobjacob) [inline contracts/campaign/Campaign.sol:71]: @cryptostaker2021 The shuttle numbers are inclusive.

File: contracts/campaign/Campaign.sol (hunk 1 above)
Comment (jacobstakeall) [reply contracts/campaign/Campaign.sol]: https://github.com/stakeall/cross-chain-staking-shuttle/blob/08f818f2cf07a0aecfc21dec9b46bc1c8a5d831f/contracts/campaign/Campaign.sol#L70

Shuttle Numbers are inclusive.

File: contracts/campaign/Campaign.sol (hunk 2 above)
Comment (jacobstakeall) [reply contracts/campaign/Campaign.sol]: https://github.com/stakeall/cross-chain-staking-shuttle/blob/08f818f2cf07a0aecfc21dec9b46bc1c8a5d831f/contracts/campaign/Campaign.sol#L69

Added.

File: contracts/campaign/Campaign.sol (hunk 3 above)
Comment (jacobstakeall) [reply contracts/campaign/Campaign.sol:36]: https://github.com/stakeall/cross-chain-staking-shuttle/blob/08f818f2cf07a0aecfc21dec9b46bc1c8a5d831f/contracts/campaign/Campaign.sol#L26

Passed childPool to the initialize

File: contracts/campaign/Campaign.sol (hunk 2 above)
Comment (jacobstakeall) [reply contracts/campaign/Campaign.sol]: https://github.com/stakeall/cross-chain-staking-shuttle/commit/3f3f66e99d67a9cda48761d300260a5172aea248

Fixed.

File: contracts/campaign/Campaign.sol (hunk 4 above)
Comment (jacobstakeall) [reply contracts/campaign/Campaign.sol:63]: This requires a discussion.

File: contracts/campaign/Campaign.sol [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+            "!Active Campaign"
+        );
+
+        require(
+            _shuttleNumber >= campaigns[_campaignNumber].startShuttleNum && 
+            _shuttleNumber <= campaigns[_campaignNumber].endShuttleNum, 
+            "!No Shuttle in Campaign"
+        );
+        
+        uint256 rewardAmount_ = campaigns[_campaignNumber].rewardAmountPerShuttle.mul(_userAmount).div(_totalAmount);
```
Comment (cryptostaker2021) [inline contracts/campaign/Campaign.sol]: You are not updating the claimed amount of that campaign? `totalClaimedAmount`

**Review by cryptostaker2021 — CHANGES_REQUESTED** [review]

**Review by robin-eth — APPROVED** [review]

**Pull request merged**